Text/Megaparsec/Stream.hs

```diff
@@ -14,284 +14,102 @@
 --
 -- @since 6.0.0
 
-{-# LANGUAGE CPP                 #-}
-{-# LANGUAGE FlexibleContexts    #-}
-{-# LANGUAGE FlexibleInstances   #-}
-{-# LANGUAGE LambdaCase          #-}
-{-# LANGUAGE MultiWayIf          #-}
-{-# LANGUAGE RankNTypes          #-}
-{-# LANGUAGE RecordWildCards     #-}
-{-# LANGUAGE ScopedTypeVariables #-}
-{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE AllowAmbiguousTypes   #-}
+{-# LANGUAGE CPP                   #-}
+{-# LANGUAGE DefaultSignatures     #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE InstanceSigs          #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE MultiWayIf            #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE RecordWildCards       #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
 
 module Text.Megaparsec.Stream
-  ( Stream (..) )
+  ( Stream (..), reachOffset, reachOffsetNoLine, takeN_, stringPretty )
 where
 
-import Data.Char (chr)
-import Data.Foldable (foldl')
-import Data.List.NonEmpty (NonEmpty (..))
-import Data.Maybe (fromMaybe)
-import Data.Proxy
-import Data.Word (Word8)
-import Text.Megaparsec.Pos
-import Text.Megaparsec.State
-import qualified Data.ByteString            as B
-import qualified Data.ByteString.Char8      as B8
-import qualified Data.ByteString.Lazy       as BL
-import qualified Data.ByteString.Lazy.Char8 as BL8
-import qualified Data.List.NonEmpty         as NE
-import qualified Data.Text                  as T
-import qualified Data.Text.Lazy             as TL
+import           Prelude               hiding (break, span, splitAt)
 
-#if !MIN_VERSION_base(4,11,0)
-import Data.Semigroup
-#endif
-
--- | Type class for inputs that can be consumed by the library.
-
-class (Ord (Token s), Ord (Tokens s)) => Stream s where
+import qualified Data.ByteString       as B
+import qualified Data.ByteString.Lazy  as BL
+import           Data.Char             (chr)
+import           Data.List.NonEmpty    (NonEmpty (..))
+import qualified Data.List.NonEmpty    as NE
+import           Data.Maybe            (fromMaybe)
+import qualified Data.Text             as T
+import qualified Data.Text.Lazy        as TL
+import           Text.Megaparsec.Pos
+import           Text.Megaparsec.State
 
-  -- | Type of token in the stream.
+import           Data.MonoTraversable
+import           Data.Sequences
 
-  type Token s :: *
+#if !MIN_VERSION_base(4,11,0)
+import           Data.Semigroup
+#endif
 
-  -- | Type of “chunk” of the stream.
 
+class (IsSequence (Tokens s), Ord (Tokens s), Ord (Element (Tokens s))) => Stream s where
   type Tokens s :: *
```

Comment:
Here's the main change. Everything else is pretty much just a consequence of this (and my hacking/messing around..)